File: Hot 100/Backtracing/0039_CombinationSum.java
//Backtracing
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();

        backtrack(candidates, ans, new ArrayList<>(), 0, target);

        return ans;
    }

    void backtrack(int[] candidates, List<List<Integer>> ans, List<Integer> path, int idx, int remain) {
        if (remain == 0) {
            ans.add(new ArrayList<>(path));
            return;
        } else if (remain < 0) return;

        for (int i = idx; i < candidates.length; i++) {
            int pick = candidates[i];

            path.add(pick);
            backtrack(candidates, ans, path, i, remain - pick);
            path.remove(path.size() - 1);
        }
    }
}

File: Top 100 Liked/Backtracking/0039_CombinationSum.py
//Backtracing
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        ans = []

        def backtrack(start: int, remain: int, path: List[int]):
            if remain == 0:
                ans.append(list(path))
                return
            if remain < 0:
                return
            for i in range(start, len(candidates)):
                pick = candidates[i]
                if pick > remain:
                    return
                path.append(pick)
                backtrack(i, remain - pick, path)
                path.pop()

        backtrack(0, target, [])

        return ans
